### PR TITLE
Fixes #3475 by checking pid when validating two lockfiles with same time

### LINKED ISSUE #3475: [rush] LockFile does not handle two processes locking in same millisecond
## Summary

If you try and acquire a LockFile from two processes within the same millisecond, then both processes acquire a dirty lock as they see each others lockfiles as stale.

## Repro steps

Will file PR with test case.

## Details

In [this block](https://github.com/microsoft/rushstack/blob/main/libraries/node-core-library/src/LockFile.ts#L339) we do not test for the case of `smallestBirthTimeMs === otherBirthtimeMs` which causes both processes to acquire valid dirty locks.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? |  5.55.0 |
| `rushVersion` from rush.json? | 5.55.0 |
| `useWorkspaces` from rush.json? | yes |
| Operating system? | Linux |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 16 |

### PR DESCRIPTION
**Pull request by benjamind**
## Summary

Under Linux or Mac, if you try and acquire a LockFile from two processes within the same millisecond, then both processes acquire a dirty lock as they see each others lockfiles as stale.

## Details

Adds a test case to catch the situation where an earlier process created a lockfile before the current process when acquiring a lock, and both processes did this within the same millisecond.

Also adds fix for this case by comparing pids in this situation and accepting the lower pid as the process which got the valid lock.

Shouldn't change any existing behaviors and should catch this edge case!

## How it was tested

Added unit tests.

### CODE REVIEW

File: libraries/node-core-library/src/test/LockFile.test.ts [hunk 1]
```diff
@@ -164,6 +164,37 @@ describe('LockFile', () => {
...
+
+        const otherPidLockFileName: string = LockFile.getLockFilePath(testFolder, resourceName, otherPid);
+
+        setLockFileGetProcessStartTime((pid: number) => {
+          return pid === process.pid ? thisPidStartTime : otherPidStartTime;
+        });
+
+        // create an open lockfile
+        const lockFileHandle: FileWriter = FileWriter.open(otherPidLockFileName);
+        lockFileHandle.write(otherPidStartTime);
```
Comment (octogonz) [inline libraries/node-core-library/src/test/LockFile.test.ts:189]: Will this test work correctly on Windows/Mac/Linux OS's?

Comment (benjamind) [reply libraries/node-core-library/src/test/LockFile.test.ts:189]: The test is within the mac/linux ifdef block above already so should be fine.

**Comment by octogonz** [comment]
Thank you for figuring this out! 🙂

File: libraries/node-core-library/src/test/LockFile.test.ts (hunk 1 above)
Comment (octogonz) [reply libraries/node-core-library/src/test/LockFile.test.ts:189]: Cool. I'll test your branch on my Windows machine just to be sure, then we can merge it. 👍

**Comment by benjamind** [comment]
> Thank you for figuring this out! 🙂

No problem, Rush is getting some heavy usage at Adobe now, so you can expect a little more engagement from some of us as we lean into it a bit more.

File: libraries/node-core-library/src/test/LockFile.test.ts (hunk 1 above)
Comment (octogonz) [reply libraries/node-core-library/src/test/LockFile.test.ts:189]: ✔ builds successfully on Windows

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
This was published with `@rushstack/node-core-library` 3.45.7

**Comment by octogonz** [comment]
> you can expect a little more engagement from some of us as we lean into it a bit more.

@benjamind if you have any specific questions or feature ideas, we do a monthly [Rush Hour](https://rushstack.io/community/events/) video call with the maintainers.

**Comment by benjamind** [comment]
Excellent thanks @octogonz!

I assume this will also go out as part of the main rush package update at some point?

**Comment by octogonz** [comment]
> I assume this will also go out as part of the main rush package update at some point?

If you need that, let us know, and we can publish a PATCH release right away. Your original request was about a use case unrelated Rush.

**Comment by benjamind** [comment]
Oh yeah no urgency on the main rush release, as we haven't yet identified that this causes a problem in general rush operations which are using the LockFile system. Though I guess it could cause some unpredictable bugs on occasion.

Thanks for getting this out so promptly!
